source/opt/dead_branch_elim_pass.cpp

```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
+            stack.push_back(get_parent_block(label));
+          });
     }
+  }
 
-    // If merge block is unreachable, continue eliminating blocks until
-    // a live block or last block is reached.
-    while (!HasNonPhiNonBackedgeRef(dLabId)) {
-      backedges_.erase((*dbi)->terminator());
-      KillAllInsts(*dbi);
-      elimBlocks.insert(*dbi);
-      ++dbi;
-      if (dbi == structuredOrder.end()) break;
-      dLabId = (*dbi)->id();
+  // Check for unreachable merge and continue blocks with live headers, those
+  // blocks must remain. Continues are tracked separately so that when updating
+  // live phi nodes with an edge from a continue I can replace it with an
+  // undef (because I still clobber the continue block).
+  std::unordered_set<ir::BasicBlock*> unreachableMerges;
+  std::unordered_map<ir::BasicBlock*, ir::BasicBlock*> unreachableContinues;
+  for (auto block : liveBlocks) {
+    uint32_t mergeId = block->MergeBlockIdIfAny();
+    if (mergeId != 0) {
+      ir::BasicBlock* mergeBlock = get_parent_block(mergeId);
+      if (!liveBlocks.count(mergeBlock)) {
+        unreachableMerges.insert(mergeBlock);
+      }
+      uint32_t contId = block->ContinueBlockIdIfAny();
+      if (contId != 0) {
+        ir::BasicBlock* contBlock = get_parent_block(contId);
+        if (!liveBlocks.count(contBlock)) {
+          unreachableContinues[contBlock] = block;
+        }
+      }
     }
+  }
 
-    // If last block reached, look for next dead branch
-    if (dbi == structuredOrder.end()) continue;
+  for (auto& block : *func) {
+    if (liveBlocks.count(&block)) {
+      // Fix phis so that only live (or unremovable) incoming edges are
+      // present. If the block now only has a single live incoming edge, remove
+      // the phi and replace its uses with its data input.
+      for (auto iter = block.begin(); iter != block.end();) {
+        if (iter->opcode() != SpvOpPhi) {
+          break;
+        }
 
-    // Create set of dead predecessors in preparation for phi update.
-    // Add the header block if the live branch is not the merge block.
-    std::unordered_set<ir::BasicBlock*> deadPreds(elimBlocks);
-    if (liveLabId != dLabId) deadPreds.insert(*bi);
+        bool changed = false;
+        ir::Instruction* inst = &*iter;
+        std::vector<ir::Operand> operands;
+        // Iterate through the incoming labels and determine which to keep
+        // and/or modify.
+        for (uint32_t i = 1; i < inst->NumInOperands(); i += 2) {
+          ir::BasicBlock* inc =
+              get_parent_block(inst->GetSingleWordInOperand(i));
+          if (unreachableContinues.count(inc)) {
+            // Replace incoming value with undef.
+            operands.emplace_back(
+                SPV_OPERAND_TYPE_ID,
+                std::initializer_list<uint32_t>{Type2Undef(inst->type_id())});
+            operands.push_back(inst->GetInOperand(i));
+            changed = true;
+          } else if (liveBlocks.count(inc)) {
```

Comment:
To fix the problem that Greg identified you need to check that the edge still exists, not just that the block is live.